File: courseschedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int>>& prerequisites) {
        vector<int> indegree(numCourses, 0);
        unordered_map<int, list<int>> adjList;
        for (auto &p: prerequisites) {
            indegree[p.first]++;
            adjList[p.second].push_front(p.first);
        }
        queue<int> que;
        int count = 0;
        for (int i = 0; i < indegree.size(); i++) 
            if (indegree[i] == 0) {
                que.push(i);
            }
        while (!que.empty()) {
            int target = que.front();
            que.pop();
            count++;
            for (auto it = adjList[target].begin(); it != adjList[target].end(); it++) {
                indegree[*it]--;
                if (indegree[*it] == 0) {
                    que.push(*it);
                }
            }
        }
        
        return count == numCourses;
    }
};

File: python/courseschedule.py
class Solution(object):
    
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        matrix = [ [] for _ in xrange(numCourses)]
        indegree = [0 for _ in xrange(numCourses)]
        for j, i in prerequisites:
            indegree[j] += 1
            matrix[i].append(j)
        count = 0
        dq = collections.deque()
        for i in xrange(numCourses):
            if indegree[i] == 0:
                dq.append(i)
        while dq:
            course = dq.popleft()
            count += 1
            for i in xrange(len(matrix[course])):
                indegree[matrix[course][i]] -= 1
                if indegree[matrix[course][i]]==0:
                    dq.append(matrix[course][i])
                        
        return count == numCourses
            
